src/storage/txn/process.rs

```diff
@@ -461,31 +460,37 @@ fn process_read_impl<E: Engine>(
     }
 }
 
-// If lock_mgr has waiters, there may be some transactions waiting for these keys,
-// so calculates keys' hashes to wake up them.
-fn gen_key_hashes_if_needed<L: LockManager, K: Borrow<Key>>(
-    lock_mgr: &Option<L>,
-    keys: &[K],
-) -> Option<Vec<u64>> {
-    lock_mgr.as_ref().and_then(|lm| {
-        if lm.has_waiter() {
-            Some(keys.iter().map(|key| key.borrow().gen_hash()).collect())
-        } else {
-            None
-        }
-    })
+#[derive(Default)]
+struct ReleasedLocks {
+    start_ts: TimeStamp,
+    commit_ts: TimeStamp,
+    hashes: Vec<u64>,
+    pessimistic: bool,
 }
 
-// Wake up pessimistic transactions that waiting for these locks
-fn wake_up_waiters_if_needed<L: LockManager>(
-    lock_mgr: &Option<L>,
-    lock_ts: TimeStamp,
-    key_hashes: Option<Vec<u64>>,
-    commit_ts: TimeStamp,
-    is_pessimistic_txn: bool,
-) {
-    if let Some(lm) = lock_mgr {
-        lm.wake_up(lock_ts, key_hashes, commit_ts, is_pessimistic_txn);
+impl ReleasedLocks {
+    fn new(start_ts: TimeStamp, commit_ts: TimeStamp) -> Self {
+        Self {
+            start_ts,
+            commit_ts,
+            ..Default::default()
+        }
+    }
+
+    fn push(&mut self, lock: ReleasedLock) {
```

Comment:
I still need to push hash to a vec to get the iter..

Reply:
So for example, rather than use

```
for k in keys {
    released_locks.push(txn.rollback(k)?);
}
```

you can use

```
released_locks.hashes(keys.iter().map(|k| txn.rollback(k)))?;
```

where `hashes` is something like:

```
fn hashes<I, T, E>(&mut self, iter: I) -> Result<(), E>
    where
        I: Iterator<Item = Result<T, E>>
{
    self.hashes = iter.collect()?;
    Ok(())
}
```

Reply:
I see. Thanks!

Reply:
I tried to implement it but I found it's more complex than `for` loop:
```rust
impl ReleasedLocks {
    fn hashes<I, E>(&mut self, iter: I) -> std::result::Result<(), E>
    where
        I: Iterator<Item = std::result::Result<Option<ReleasedLock>, E>>,
    {
        self.hashes = iter
            .filter_map(|v| match v {
                Ok(Some(lock)) => {
                    if !self.pessimistic {
                        self.pessimistic = lock.pessimistic;
                    }
                    Some(Ok(lock.hash))
                }
                Ok(None) => None,
                Err(e) => Some(Err(e)),
            })
            .collect::<std::result::Result<Vec<u64>, E>>()?;
        Ok(())
    }
}

fn process_write_impl(...) {
...
            let mut released_locks = ReleasedLocks::new(lock_ts, commit_ts);
            // for k in keys {
            // released_locks.push(txn.commit(k, commit_ts)?);
            // }
            released_locks.hashes(keys.into_iter().map(|k| txn.commit(k, commit_ts)))?;
            released_locks.wake_up(lock_mgr.as_ref());
...
}
```

Reply:
And I still need to call `push` in `ResolveLock`..

Reply:
@nrc PTAL again. Thanks!